Hashes/MD5.js

```diff
@@ -0,0 +1,221 @@
+ * @return {string} - string representation padding with empty (0) bits
+ *
+ * @example
+ *      pad("10011", 8); // "00010011"
+ */
+function pad(str, bits) {
+  let res = str
+  while (res.length % bits !== 0) {
+    res = '0' + res
+  }
+  return res
+}
+
+/**
+ * Separates string into chunks of the same size
+ *
+ * @param {string} str - string to separate into chunks
+ * @param {int} size - number of characters wanted in each chunk
+ * @return {array} - array of original string split into chunks
+ *
+ * @example
+ *      chunkify("this is a test", 2)
+ */
+function chunkify(str, size) {
+  const chunks = []
+  for (let i = 0; i < str.length; i += size) {
+    chunks.push(str.slice(i, i + size))
+  }
+  return chunks
+}
+
+/**
+ * Rotates the bits to the left
+ *
+ * @param {number} bits - 32 bit number
+ * @param {int} turns - number of rotations to make
+ * @return {string} - number after bits rotation
+ *
+ * @example
+ *      rotateLeft("1011", 3); // "1101"
+ */
+function rotateLeft(bits, turns) {
+  return (bits << turns) | (bits >>> (32 - turns))
+}
+
+/**
+ * Pre-processes message to feed the algorithm loop
+ *
+ * @param {string} message - message to pre-process
+ * @return {string} - processed message
+ */
+function preProcess(message) {
+  // convert message to binary representation padded to
+  // 8 bits, and add 1
+  let m =
+    message
+      .split('')
+      .map((e) => e.charCodeAt(0))
+      .map((e) => e.toString(2))
+      .map((e) => pad(e, 8))
+      .join('') + '1'
+
+  // extend message by adding empty bits (0)
+  while (m.length % 512 !== 448) {
+    m += '0'
+  }
+
+  m = chunkify(m, 32)
+    .map((e) => toLittleEndian(e))
+    .join('')
+
+  // length of message in binary, padded, and extended
+  // to a 64 bit representation
+  let ml = (message.length * CHAR_SIZE).toString(2)
+  ml = pad(ml, 64)
+
+  return m + ml.slice(32) + ml.slice(0, 32)
+}
+
+/**
+ * Hashes message using MD5 Cryptographic Hash Function
+ *
+ * @see
+ * For more info: https://en.wikipedia.org/wiki/MD5
+ * 
+ * @param {string} message - message to hash
+ * @return {string} - message digest (hash value)
+ */
+function MD5(message) {
+  // Initialize variables:
+  const [a0, b0, c0, d0] = [0, 1, 2, 3]
+  const hashes = Uint32Array.from([
+    0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476
+  ])
+
+  // pre-process message and split into 512 bit chunks
+  const bits = preProcess(message)
+  const chunks = chunkify(bits, 512)
+
+  chunks.forEach(function (chunk, _) {
+    // break each chunk into 16 32-bit words
+    const words = Uint32Array.from(
+      chunkify(chunk, 32).map((e) => parseInt(e, 2))
+    )
+
+    // initialize variables for this chunk
+    const [A, B, C, D] = [0, 1, 2, 3]
+    const abcd = Uint32Array.from([
+      hashes[a0],
+      hashes[b0],
+      hashes[c0],
+      hashes[d0]
+    ])
+
+    for (let i = 0; i < 64; i++) {
+      const [F, g] = [0, 1]
+      const fg = Uint32Array.from([0, 0])
```

Comment:
An Uint32Array for two variables? Why?

Reply:
Uint32 Arrays were used to avoid manual wrapping addition, and (type casting?) using `num >>> 0`.
By using typed arrays we ensure that the numbers stored are 32 bit unsigned integers. I opted for typed arrays as an alternative to using the >>> operator due to my lack of confidence in its usage.